File: curriculum/15_metrics_collector/go-impl/metrics/store.go
package metrics

import (
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"sort"
	"strings"
	"sync"
	"time"
)

type MetricType string

const (
	TypeCounter  MetricType = "counter"
	TypeGauge    MetricType = "gauge"
	TypeHistogram MetricType = "histogram"
	TypeTimer    MetricType = "timer"
)

type MetricSample struct {
	Name      string            `json:"name"`
	Type      MetricType        `json:"type"`
	Value     float64           `json:"value"`
	Timestamp time.Time         `json:"timestamp"`
	Labels    map[string]string `json:"labels"`
}

type MetricSeries struct {
	Name   string
	Type   MetricType
	Labels map[string]string
}

func (s MetricSeries) Key() string {
	var parts []string
	parts = append(parts, string(s.Type), s.Name)
	var keys []string
	for k := range s.Labels {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	for _, k := range keys {
		parts = append(parts, k+"="+s.Labels[k])
	}
	return strings.Join(parts, ",")
}

type TimeSeriesPoint struct {
	Timestamp time.Time `json:"timestamp"`
	Value     float64   `json:"value"`
}

type HistogramBucket struct {
	UpperBound      float64 `json:"upper_bound"`
	CumulativeCount int64   `json:"cumulative_count"`
}

type HistogramData struct {
	Buckets []HistogramBucket `json:"buckets"`
	Count   int64             `json:"count"`
	Sum     float64           `json:"sum"`
}

type AlertRule struct {
	RuleID    string  `json:"rule_id"`
	Name      string  `json:"name"`
	Enabled   bool    `json:"enabled"`
	Query     string  `json:"query"`
	Operator  string  `json:"operator"`
	Threshold float64 `json:"threshold"`
	WindowSeconds int `json:"window_seconds"`
	Severity  string  `json:"severity"`
}

type AlertEvent struct {
	AlertEventID string    `json:"alert_event_id"`
	RuleID       string    `json:"rule_id"`
	TriggeredAt  time.Time `json:"triggered_at"`
	ObservedValue float64  `json:"observed_value"`
	Threshold    float64   `json:"threshold"`
	Severity     string    `json:"severity"`
}

type MetricStore struct {
	mu         sync.RWMutex
	samples    map[string][]TimeSeriesPoint
	histograms map[string]*HistogramData
	alerts     map[string]*AlertRule
	events     []AlertEvent
	maxSize    int
}

func NewMetricStore(maxSize int) *MetricStore {
	return &MetricStore{
		samples:    make(map[string][]TimeSeriesPoint),
		histograms: make(map[string]*HistogramData),
		alerts:     make(map[string]*AlertRule),
		events:     make([]AlertEvent, 0),
		maxSize:    maxSize,
	}
}

func (s *MetricStore) Record(sample MetricSample) {
	s.mu.Lock()
	defer s.mu.Unlock()

	key := MetricSeries{Name: sample.Name, Type: sample.Type, Labels: sample.Labels}.Key()

	switch sample.Type {
	case TypeCounter, TypeGauge:
		points := s.samples[key]
		if len(points) >= s.maxSize {
			points = points[1:]
		}
		points = append(points, TimeSeriesPoint{Timestamp: sample.Timestamp, Value: sample.Value})
		s.samples[key] = points
	case TypeHistogram, TypeTimer:
		h := s.histograms[key]
		if h == nil {
			h = &HistogramData{
				Buckets: []HistogramBucket{
					{UpperBound: 0.005}, {UpperBound: 0.01}, {UpperBound: 0.025},
					{UpperBound: 0.05}, {UpperBound: 0.1}, {UpperBound: 0.25},
					{UpperBound: 0.5}, {UpperBound: 1}, {UpperBound: 2.5},
					{UpperBound: 5}, {UpperBound: 10}, {UpperBound: float64(int64(^uint64(0) >> 1))},
				},
			}
			s.histograms[key] = h
		}
		h.Count++
		h.Sum += sample.Value
		for i := range h.Buckets {
			if sample.Value <= h.Buckets[i].UpperBound {
				h.Buckets[i].CumulativeCount++
			}
		}
	}
}

func (s *MetricStore) Query(name string, metricType MetricType, labels map[string]string, start, end time.Time, aggregation string) float64 {
	s.mu.RLock()
	defer s.mu.RUnlock()

	key := MetricSeries{Name: name, Type: metricType, Labels: labels}.Key()
	points := s.samples[key]

	var values []float64
	for _, p := range points {
		if !start.IsZero() && p.Timestamp.Before(start) {
			continue
		}
		if !end.IsZero() && p.Timestamp.After(end) {
			continue
		}
		values = append(values, p.Value)
	}

	return s.aggregate(values, aggregation)
}

func (s *MetricStore) aggregate(values []float64, aggregation string) float64 {
	if len(values) == 0 {
		return 0
	}

	switch aggregation {
	case "sum":
		var sum float64
		for _, v := range values {
			sum += v
		}
		return sum
	case "avg":
		var sum float64
		for _, v := range values {
			sum += v
		}
		return sum / float64(len(values))
	case "min":
		min := values[0]
		for _, v := range values {
			if v < min {
				min = v
			}
		}
		return min
	case "max":
		max := values[0]
		for _, v := range values {
			if v > max {
				max = v
			}
		}
		return max
	case "count":
		return float64(len(values))
	case "p50", "p95", "p99":
		sort.Float64s(values)
		idx := 0
		switch aggregation {
		case "p50":
			idx = int(float64(len(values)-1) * 0.50)
		case "p95":
			idx = int(float64(len(values)-1) * 0.95)
		case "p99":
			idx = int(float64(len(values)-1) * 0.99)
		}
		return values[idx]
	}
	return 0
}

func (s *MetricStore) HistogramPercentile(name string, labels map[string]string, percentile float64) float64 {
	s.mu.RLock()
	defer s.mu.RUnlock()

	key := MetricSeries{Name: name, Type: TypeHistogram, Labels: labels}.Key()
	h := s.histograms[key]
	if h == nil || h.Count == 0 {
		key = MetricSeries{Name: name, Type: TypeTimer, Labels: labels}.Key()
		h = s.histograms[key]
	}
	if h == nil || h.Count == 0 {
		return 0
	}

	target := int64(float64(h.Count) * percentile)
	for _, b := range h.Buckets {
		if b.CumulativeCount >= target {
			return b.UpperBound
		}
	}
	return 0
}

func (s *MetricStore) CreateAlert(rule AlertRule) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.alerts[rule.RuleID] = &rule
}

func (s *MetricStore) EvaluateAlerts() {
	s.mu.Lock()
	defer s.mu.Unlock()

	for _, rule := range s.alerts {
		if !rule.Enabled {
			continue
		}
		// Simplified: evaluate against counter/gauge samples
		parts := strings.Split(rule.Query, "(")
		if len(parts) < 2 {
			continue
		}
		agg := strings.TrimSpace(parts[0])
		name := strings.TrimSuffix(strings.TrimSpace(parts[1]), ")")

		value := s.aggregateQuery(name, agg, time.Now().Add(-time.Duration(rule.WindowSeconds)*time.Second), time.Now())
		triggered := false
		switch rule.Operator {
		case "gt":
			triggered = value > rule.Threshold
		case "gte":
			triggered = value >= rule.Threshold
		case "lt":
			triggered = value < rule.Threshold
		case "lte":
			triggered = value <= rule.Threshold
		}

		if triggered {
			s.events = append(s.events, AlertEvent{
				AlertEventID:  fmt.Sprintf("evt_%d", time.Now().UnixNano()),
				RuleID:        rule.RuleID,
				TriggeredAt:   time.Now(),
				ObservedValue: value,
				Threshold:     rule.Threshold,
				Severity:      rule.Severity,
			})
		}
	}
}

func (s *MetricStore) aggregateQuery(name, aggregation string, start, end time.Time) float64 {
	var allValues []float64
	for key, points := range s.samples {
		if strings.Contains(key, name) {
			for _, p := range points {
				if p.Timestamp.After(start) && p.Timestamp.Before(end) {
					allValues = append(allValues, p.Value)
				}
			}
		}
	}
	return s.aggregate(allValues, aggregation)
}

func (s *MetricStore) PrometheusExport() string {
	s.mu.RLock()
	defer s.mu.RUnlock()

	var output strings.Builder
	for key, points := range s.samples {
		parts := strings.SplitN(key, ",", 3)
		if len(parts) < 2 {
			continue
		}
		name := parts[1]
		value := 0.0
		if len(points) > 0 {
			value = points[len(points)-1].Value
		}
		output.WriteString(fmt.Sprintf("# TYPE %s %s\n", name, parts[0]))
		output.WriteString(fmt.Sprintf("%s %s %g\n", name, formatLabels(parts[2:]), value))
	}

	for key, h := range s.histograms {
		parts := strings.SplitN(key, ",", 3)
		if len(parts) < 2 {
			continue
		}
		name := parts[1]
		output.WriteString(fmt.Sprintf("# TYPE %s histogram\n", name))
		for _, b := range h.Buckets {
			output.WriteString(fmt.Sprintf("%s_bucket{le=\"%g\"} %d\n", name, b.UpperBound, b.CumulativeCount))
		}
		output.WriteString(fmt.Sprintf("%s_sum %g\n", name, h.Sum))
		output.WriteString(fmt.Sprintf("%s_count %d\n", name, h.Count))
	}

	return output.String()
}

func formatLabels(parts []string) string {
	if len(parts) == 0 || parts[0] == "" {
		return ""
	}
	var labels []string
	for _, p := range parts {
		labels = append(labels, p)
	}
	return "{" + strings.Join(labels, ",") + "}"
}

type Server struct {
	store *MetricStore
}

func NewServer() *Server {
	return &Server{store: NewMetricStore(10000)}
}

func (s *Server) ServeHTTP(w http.ResponseWriter, r *http.Request) {
	switch r.URL.Path {
	case "/metrics/counter", "/metrics/gauge", "/metrics/histogram", "/metrics/timer":
		if r.Method == http.MethodPost {
			s.handleRecord(w, r)
		} else {
			http.Error(w, "method not allowed", http.StatusMethodNotAllowed)
		}
	case "/metrics":
		if r.Method == http.MethodGet {
			if r.URL.Query().Get("query") != "" {
				s.handleQuery(w, r)
			} else {
				s.handlePrometheus(w, r)
			}
		} else {
			http.Error(w, "method not allowed", http.StatusMethodNotAllowed)
		}
	case "/dashboard":
		if r.Method == http.MethodGet {
			s.handleDashboard(w, r)
		} else {
			http.Error(w, "method not allowed", http.StatusMethodNotAllowed)
		}
	case "/alerts/rules":
		if r.Method == http.MethodPost {
			s.handleCreateAlert(w, r)
		} else if r.Method == http.MethodGet {
			s.handleListAlerts(w, r)
		} else {
			http.Error(w, "method not allowed", http.StatusMethodNotAllowed)
		}
	case "/health":
		s.handleHealth(w, r)
	default:
		http.Error(w, "not found", http.StatusNotFound)
	}
}

func (s *Server) handleRecord(w http.ResponseWriter, r *http.Request) {
	var sample MetricSample
	if err := json.NewDecoder(r.Body).Decode(&sample); err != nil {
		s.writeError(w, http.StatusBadRequest, "invalid_metric_sample", err.Error())
		return
	}
	if sample.Name == "" {
		s.writeError(w, http.StatusBadRequest, "invalid_metric_sample", "name is required")
		return
	}
	if sample.Timestamp.IsZero() {
		sample.Timestamp = time.Now()
	}

	metricType := MetricType(strings.TrimPrefix(r.URL.Path, "/metrics/"))
	sample.Type = metricType

	s.store.Record(sample)
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(http.StatusAccepted)
	json.NewEncoder(w).Encode(map[string]any{
		"ok": true,
		"data": map[string]any{"accepted": 1, "duplicates": 0, "rejected": 0},
	})
}

func (s *Server) handleQuery(w http.ResponseWriter, r *http.Request) {
	query := r.URL.Query().Get("query")
	startStr := r.URL.Query().Get("start")
	endStr := r.URL.Query().Get("end")

	var start, end time.Time
	if startStr != "" {
		start, _ = time.Parse(time.RFC3339, startStr)
	}
	if endStr != "" {
		end, _ = time.Parse(time.RFC3339, endStr)
	}

	// Parse simple query: aggregation(name)
	agg, name := parseQuery(query)
	value := s.store.Query(name, TypeGauge, nil, start, end, agg)

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(map[string]any{
		"ok": true,
		"data": map[string]any{
			"query": query,
			"value": value,
		},
	})
}

func (s *Server) handlePrometheus(w http.ResponseWriter, _ *http.Request) {
	w.Header().Set("Content-Type", "text/plain; version=0.0.4")
	w.Write([]byte(s.store.PrometheusExport()))
}

func (s *Server) handleDashboard(w http.ResponseWriter, r *http.Request) {
	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(map[string]any{
		"ok": true,
		"data": map[string]any{
			"dashboard_id": "default",
			"panels":       []any{},
			"alerts":       []any{},
		},
	})
}

func (s *Server) handleCreateAlert(w http.ResponseWriter, r *http.Request) {
	var rule AlertRule
	if err := json.NewDecoder(r.Body).Decode(&rule); err != nil {
		s.writeError(w, http.StatusBadRequest, "invalid_alert_rule", err.Error())
		return
	}
	s.store.CreateAlert(rule)
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(http.StatusCreated)
	json.NewEncoder(w).Encode(map[string]any{
		"ok": true,
		"data": map[string]any{"rule_id": rule.RuleID, "status": "enabled"},
	})
}

func (s *Server) handleListAlerts(w http.ResponseWriter, _ *http.Request) {
	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(map[string]any{
		"ok": true,
		"data": map[string]any{"items": []any{}},
	})
}

func (s *Server) handleHealth(w http.ResponseWriter, _ *http.Request) {
	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(map[string]any{
		"ok": true,
		"data": map[string]any{
			"status":         "ok",
			"durability_mode": "volatile_until_flush",
			"active_series":   0,
		},
	})
}

func (s *Server) writeError(w http.ResponseWriter, status int, code, message string) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(status)
	json.NewEncoder(w).Encode(map[string]any{
		"ok": false,
		"error": map[string]any{
			"code":    code,
			"message": message,
		},
	})
}

func parseQuery(query string) (string, string) {
	// Simple parser for aggregation(name)
	idx := strings.Index(query, "(")
	if idx == -1 {
		return "sum", query
	}
	agg := strings.TrimSpace(query[:idx])
	name := strings.TrimSuffix(strings.TrimSpace(query[idx+1:]), ")")
	return agg, name
}

func NewShutdownContext(timeout time.Duration) (context.Context, context.CancelFunc) {
	return context.WithTimeout(context.Background(), timeout)
}

File: curriculum/15_metrics_collector/node-impl/src/store.ts
export type MetricType = 'counter' | 'gauge' | 'histogram' | 'timer';

export interface MetricSample {
  name: string;
  type: MetricType;
  value: number;
  timestamp: Date;
  labels: Record<string, string>;
}

export interface TimeSeriesPoint {
  timestamp: Date;
  value: number;
}

export interface HistogramBucket {
  upperBound: number;
  cumulativeCount: number;
}

export interface HistogramData {
  buckets: HistogramBucket[];
  count: number;
  sum: number;
}

export interface AlertRule {
  ruleId: string;
  name: string;
  enabled: boolean;
  query: string;
  operator: string;
  threshold: number;
  windowSeconds: number;
  severity: string;
}

export interface AlertEvent {
  alertEventId: string;
  ruleId: string;
  triggeredAt: Date;
  observedValue: number;
  threshold: number;
  severity: string;
}

export interface AlertState {
  ruleId: string;
  name: string;
  status: 'ok' | 'firing';
  currentValue: number | null;
  threshold: number;
  lastEvaluatedAt: Date;
}

export interface SeriesRef {
  name: string;
  type: MetricType;
  labels: Record<string, string>;
}

export interface QueryResult {
  matched: boolean;
  type: MetricType | null;
  value: number | null;
}

function seriesKey(name: string, type: MetricType, labels: Record<string, string>): string {
  const labelParts = Object.entries(labels)
    .sort(([a], [b]) => a.localeCompare(b))
    .map(([k, v]) => `${k}=${v}`);
  return [type, name, ...labelParts].join(',');
}

function aggregate(values: number[], aggregation: string): number {
  if (values.length === 0) return 0;

  switch (aggregation) {
    case 'sum':
      return values.reduce((a, b) => a + b, 0);
    case 'avg':
      return values.reduce((a, b) => a + b, 0) / values.length;
    case 'min':
      return Math.min(...values);
    case 'max':
      return Math.max(...values);
    case 'count':
      return values.length;
    case 'p50':
    case 'p95':
    case 'p99': {
      const sorted = [...values].sort((a, b) => a - b);
      const pct = aggregation === 'p50' ? 0.50 : aggregation === 'p95' ? 0.95 : 0.99;
      const idx = Math.floor((sorted.length - 1) * pct);
      return sorted[idx];
    }
    default:
      return 0;
  }
}

function histogramAggregate(h: HistogramData, aggregation: string): number | null {
  switch (aggregation) {
    case 'count':
      return h.count;
    case 'sum':
      return h.sum;
    case 'avg':
      return h.count === 0 ? null : h.sum / h.count;
    case 'p50':
    case 'p95':
    case 'p99': {
      if (h.count === 0) return null;
      const pct = aggregation === 'p50' ? 0.5 : aggregation === 'p95' ? 0.95 : 0.99;
      const target = Math.ceil(h.count * pct);
      for (const bucket of h.buckets) {
        if (bucket.upperBound === Number.MAX_VALUE) break;
        if (bucket.cumulativeCount >= target) return bucket.upperBound;
      }
      return null;
    }
    default:
      return null;
  }
}

export class MetricStore {
  private samples: Map<string, TimeSeriesPoint[]> = new Map();
  private histograms: Map<string, HistogramData> = new Map();
  private alerts: Map<string, AlertRule> = new Map();
  private alertStates: Map<string, AlertState> = new Map();
  private seriesRegistry: Map<string, SeriesRef> = new Map();
  public events: AlertEvent[] = [];

  constructor(private maxSize: number) {}

  record(sample: MetricSample): void {
    const key = seriesKey(sample.name, sample.type, sample.labels);
    if (!this.seriesRegistry.has(key)) {
      this.seriesRegistry.set(key, { name: sample.name, type: sample.type, labels: sample.labels });
    }

    switch (sample.type) {
      case 'counter':
      case 'gauge': {
        let points = this.samples.get(key) ?? [];
        if (points.length >= this.maxSize) {
          points = points.slice(1);
        }
        points.push({ timestamp: sample.timestamp, value: sample.value });
        this.samples.set(key, points);
        break;
      }
      case 'histogram':
      case 'timer': {
        let h = this.histograms.get(key);
        if (!h) {
          h = {
            buckets: [
              0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1, 2.5, 5, 10, Number.MAX_VALUE,
            ].map((b) => ({ upperBound: b, cumulativeCount: 0 })),
            count: 0,
            sum: 0,
          };
          this.histograms.set(key, h);
        }
        h.count++;
        h.sum += sample.value;
        for (const bucket of h.buckets) {
          if (sample.value <= bucket.upperBound) {
            bucket.cumulativeCount++;
          }
        }
        break;
      }
    }
  }

  query(
    name: string,
    metricType: MetricType,
    labels: Record<string, string>,
    start: Date | undefined,
    end: Date | undefined,
    aggregation: string,
  ): number {
    const key = seriesKey(name, metricType, labels);
    const points = this.samples.get(key) ?? [];

    const values = points
      .filter((p) => (!start || p.timestamp >= start) && (!end || p.timestamp <= end))
      .map((p) => p.value);

    return aggregate(values, aggregation);
  }

  queryByName(
    name: string,
    labels: Record<string, string>,
    start: Date | undefined,
    end: Date | undefined,
    aggregation: string,
  ): QueryResult {
    for (const metricType of ['counter', 'gauge'] as const) {
      const key = seriesKey(name, metricType, labels);
      const points = this.samples.get(key);
      if (points && points.length > 0) {
        return {
          matched: true,
          type: metricType,
          value: this.query(name, metricType, labels, start, end, aggregation),
        };
      }
    }
    for (const metricType of ['histogram', 'timer'] as const) {
      const key = seriesKey(name, metricType, labels);
      const h = this.histograms.get(key);
      if (h && h.count > 0) {
        return { matched: true, type: metricType, value: histogramAggregate(h, aggregation) };
      }
    }
    return { matched: false, type: null, value: null };
  }

  listSeries(): SeriesRef[] {
    return [...this.seriesRegistry.values()];
  }

  seriesPoints(
    name: string,
    metricType: MetricType,
    labels: Record<string, string>,
    start: Date | undefined,
    end: Date | undefined,
  ): TimeSeriesPoint[] {
    const key = seriesKey(name, metricType, labels);
    const points = this.samples.get(key) ?? [];
    return points.filter((p) => (!start || p.timestamp >= start) && (!end || p.timestamp <= end));
  }

  histogramFor(name: string, metricType: MetricType, labels: Record<string, string>): HistogramData | undefined {
    return this.histograms.get(seriesKey(name, metricType, labels));
  }

  activeSeriesCount(): number {
    return this.seriesRegistry.size;
  }

  histogramPercentile(name: string, labels: Record<string, string>, percentile: number): number {
    const key = seriesKey(name, 'histogram', labels);
    let h = this.histograms.get(key);
    if (!h || h.count === 0) {
      const timerKey = seriesKey(name, 'timer', labels);
      h = this.histograms.get(timerKey);
    }
    if (!h || h.count === 0) return 0;

    const target = Math.ceil(h.count * percentile);
    for (const bucket of h.buckets) {
      if (bucket.cumulativeCount >= target) {
        return bucket.upperBound;
      }
    }
    return 0;
  }

  createAlert(rule: AlertRule): void {
    this.alerts.set(rule.ruleId, rule);
  }

  listAlerts(): AlertRule[] {
    return [...this.alerts.values()];
  }

  listAlertStates(): AlertState[] {
    return [...this.alertStates.values()];
  }

  listAlertEvents(ruleId?: string, limit = 100): AlertEvent[] {
    const filtered = ruleId ? this.events.filter((e) => e.ruleId === ruleId) : this.events;
    return filtered.slice(-limit);
  }

  evaluateAlerts(): void {
    for (const rule of this.alerts.values()) {
      if (!rule.enabled) continue;

      const match = rule.query.match(/^(\w+)\((\w+)\)$/);
      if (!match) continue;
      const [, agg, name] = match;

      const now = new Date();
      const start = new Date(now.getTime() - rule.windowSeconds * 1000);
      const value = this.aggregateQuery(name, agg, start, now);

      let triggered = false;
      switch (rule.operator) {
        case 'gt':
          triggered = value > rule.threshold;
          break;
        case 'gte':
          triggered = value >= rule.threshold;
          break;
        case 'lt':
          triggered = value < rule.threshold;
          break;
        case 'lte':
          triggered = value <= rule.threshold;
          break;
      }

      this.alertStates.set(rule.ruleId, {
        ruleId: rule.ruleId,
        name: rule.name,
        status: triggered ? 'firing' : 'ok',
        currentValue: value,
        threshold: rule.threshold,
        lastEvaluatedAt: now,
      });

      if (triggered) {
        this.events.push({
          alertEventId: `evt_${Date.now()}_${this.events.length}`,
          ruleId: rule.ruleId,
          triggeredAt: new Date(),
          observedValue: value,
          threshold: rule.threshold,
          severity: rule.severity,
        });
      }
    }
  }

  private aggregateQuery(name: string, aggregation: string, start: Date, end: Date): number {
    const values: number[] = [];
    for (const [key, points] of this.samples) {
      if (key.includes(name)) {
        for (const p of points) {
          if (p.timestamp >= start && p.timestamp <= end) {
            values.push(p.value);
          }
        }
      }
    }
    return aggregate(values, aggregation);
  }

  prometheusExport(): string {
    const lines: string[] = [];

    for (const [key, points] of this.samples) {
      const parts = key.split(',');
      if (parts.length < 2) continue;
      const name = parts[1];
      const value = points.length > 0 ? points[points.length - 1].value : 0;
      lines.push(`# TYPE ${name} ${parts[0]}`);
      const labelStr = parts.length > 2 && parts[2] ? `{${parts.slice(2).join(',')}}` : '';
      lines.push(`${name}${labelStr} ${value}`);
    }

    for (const [key, h] of this.histograms) {
      const parts = key.split(',');
      if (parts.length < 2) continue;
      const name = parts[1];
      lines.push(`# TYPE ${name} histogram`);
      for (const bucket of h.buckets) {
        lines.push(`${name}_bucket{le="${bucket.upperBound}"} ${bucket.cumulativeCount}`);
      }
      lines.push(`${name}_sum ${h.sum}`);
      lines.push(`${name}_count ${h.count}`);
    }

    return lines.join('\n');
  }
}
